### ISSUE TITLE
span / sub naming is confusing

### ISSUE THREAD

**Issue by sherief** [biojppm/rapidyaml#7]
I've been trying to introduce other people to the API, and the span concept was a bit confusing at first, but not too bad. The renamed sub and csub though seem to be a regression. 

While from our POV they might seem obvious, the (admittedly anecdotal) first question I got was "sub what?" - is it sub as in below something? Subscription? Even the C++ STL didn't go shorter than substr.

Looking at the STL for inspiration, why can't spans be replaced with std::string_view<> if C++17 is an acceptable target, or a custom ryml::string_view if previous C++ standards compatibility is required? The name is a lot more obvious, not too verbose, and we have `using` declarations now. I think this would make it a lot easier for someone new to the API to understand what a piece of code is doing without having to read comments or headers, which is always A Good Thing.

**Comment by biojppm** [comment]
I know. Here's the thing. 

STL's `string_view` is not writable, which this class emphatically _is._ So `string_view` is just not a good name.

My original name was span, but now or presumably soon there will be also an STL span, but this is merely a linear non-owning container of general-typed objects - definitely not a string type. So using span would be also misleading, as this class is a string-type with string methods, which no std span will ever have. So span cannot be the name, either.

So we need to choose a name other than string view or span; the name must allude to the object being a string type, and furthermore must make it known that we're dealing with a non-zero-terminated, non-owning, writable string(or not, depending on the 'c' prefix). A substring.

So I went for `subs`. I could have gone for `substr`, or less likely with `substring`. I'm always for shorter names and less typing (though I do acknowledge that conciseness must compromise with clarity). But the reasoning is this: yes, it's strange and terse, but you take the pill once (even though it may be bitter), find out what it's about and see that it's an instance of the `basic_substring<>` template. Then you immediately see that it is a substring, and from there on you can revel in the short names and faster typing and reading which its name enables. That was something I loved in `span` and tried to retain in the new name. 

OTOH, `substr` would be only two more characters, and would raise less questions. I admit I'm on the fence here.

**Comment by sherief** [comment]
You're absolutely correct, if it doesn't behave like a string_view it shouldn't be named like one.
I've been thinking about this for a few hours and something just hit me. First I had some generic names like "buffer", but that doesn't convey mutability, then "buffer_ref", and some others, but I've just been hit with this:
`range`.
The region you're parsing is defined by a begin and one-past-end pointer in memory - that's all the span is [pointer + size encoding is basically a different way of looking at the same]. It also plays nicely with the concept of a range in C++ as seen in range-based for loops and the like.
How does `ryml::parse(range(string, string+length));` sound?

**Comment by biojppm** [comment]
Do you know there's a ranges library coming to the STL? It's quite nice actually.

But anyway, it's important here to let the user know that it's a string, with string methods. And a range as it will be understood is like span; it refers to a collection of general typed objects. So I don't think `range` is a good name either.

Having said that, I recognize subs is maybe too cryptic. I'll compromise and go with `substr`. It's short enough, and clear enough. When I have opportunity, I'll change the names.

**Comment by sherief** [comment]
Can I haggle you into making it substr_ref, to differentiate it from std::string::substr() which makes a copy?

**Comment by biojppm** [comment]
Sorry, too long... And std::string::substr() is a method, so there's not much chance of confusion. In fact, if you look at it this way, a substr existing alone immediately implies non-ownership by suggesting that its source data is elsewhere (if it's a 'sub' something, then that something surely exists).
